XS1 component

markusressel opened the pull request:
## Description:

Added support for the [EZcontrol XS1 Gateway](http://www.ezcontrol.de/content/view/36/28/).
As documented in home-assistant.io this version only covers part of the capabilities of the gateway but it should be a good starting point (covers all the things I needed for my personal usage)

This is my first time opening a PR on home-assistant. I put a lot of hours into understanding how things work in hass and I hope I managed to get a lot of things right but if there is anything that should be improved please let me know :smile: 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7771

## Example entry for `configuration.yaml` (if applicable):
See documentation PR

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works. (no tests were added)

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

markusressel committed:
added xs1 main component
added implementations for switch, sensor, climate and binary_sensor

markusressel committed:
updated code
fixed styling
added comments
removed binary_sensor (wasn't working)

markusressel committed:
ran "gen_requirements_all.py" script

markusressel committed:
fixed linting issues

markusressel committed:
added config options for port and ssl
small fixes

markusressel committed:
use already defined config constants instead of defining new ones

markusressel committed:
avoid passing in hass to the entity

markusressel committed:
use async keyword and proper asyncio calls
limit updates with a global lock to prevent overwhelming the gateway with concurrent requests
change info logger calls to debug

markusressel committed:
update dependency

markusressel committed:
removed unneeded constant

markusressel committed:
fix lint issues

markusressel committed:
updated requirements

markusressel committed:
removed unused imports

markusressel commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,62 @@
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/xs1/
+"""
+import asyncio
```
fixed

markusressel commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,52 @@
+https://home-assistant.io/components/xs1/
+"""
+
+import asyncio
```
fixed

markusressel committed:
fixed some flake8 errors

markusressel commented:
I want to mention that I got some flake8 errors I didn't get when I ran the `script/lint` script.
Is this a known issue? Or did I forget to do something when executing it?

```
./homeassistant/components/xs1.py:64:1: D400 First line should end with a period
./homeassistant/components/xs1.py:110:1: D202 No blank lines allowed after function docstring
./homeassistant/components/climate/xs1.py:21:1: D401 First line should be in imperative mood; try rephrasing
./homeassistant/components/climate/xs1.py:77:1: D401 First line should be in imperative mood; try rephrasing
./homeassistant/components/climate/xs1.py:82:1: D401 First line should be in imperative mood
./homeassistant/components/climate/xs1.py:114:1: D400 First line should end with a period
./homeassistant/components/switch/xs1.py:18:1: D401 First line should be in imperative mood; try rephrasing
./homeassistant/components/sensor/xs1.py:18:1: D401 First line should be in imperative mood; try rephrasing
./homeassistant/components/sensor/xs1.py:18:1: D202 No blank lines allowed after function docstring
```

markusressel committed:
fixed some flake8 errors

markusressel committed:
changed imports to absolute paths

markusressel committed:
fixed some lint errors

markusressel committed:
fixed some lint errors

markusressel committed:
fix update of attached sensor

fabaff commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,120 @@
+    TEMP_CELSIUS)
+from homeassistant.util.temperature import convert as convert_temperature
+from homeassistant.components.xs1 import ACTUATORS, DOMAIN, SENSORS, \
+    XS1DeviceEntity
```
Import should be ordered. Use `isort` if you don't want to do it manually.

markusressel replied:
done

fabaff commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+UPDATE_LOCK = asyncio.Lock()
+
+
+def _create_controller_api(host, port, ssl, user, password):
```
Docstring is missing.

markusressel replied:
Added

fabaff commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+SENSORS = 'sensors'
+
+# configuration keys
+CONF_SSL = 'ssl'
```
Import that from `const.py`.

markusressel replied:
Oh, didnt't think this was in const. Fixed!

fabaff commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+
+    xs1 = await hass.async_add_executor_job(
+        partial(_create_controller_api,
+                host, port, ssl, user, password))
```
You should check the credentials. If they are not valid then the component setup should fail with a log entry.

Also, it could be that the `host` is offline. This could be handled with `PlatformNotReady`.

markusressel replied:
I'm not exactly sure how easy it is to check credentials, I never used them on my device. I will try to check on that later.

markusressel replied:
Ok so I checked again with my gateway and I simply cannot use the password feature (seems like a firmware bug which prevents me from setting a password). I guess that's why I never used authentication. I don't think this is a severe issue though as I don't think many people still use this gateway and it can be isolated in the network pretty easily as it doesn't need to communicate with anything else other than home assistant over the network.

If someone else has a gateway with a non default password I am happy to include code that validates credentials on initialization but I just can't test this atm :'(

fabaff commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+        self.device = device
+
+    async def async_update(self):
+        """Fetch new state data for the sensor.
```
Stale docstring.

markusressel replied:
I don't exactly know what that is supposed to mean as a change request but I aligned it with what I found in other components.

fabaff commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT): cv.string,
```
If there is a standard port then that port should be the default.

markusressel replied:
ok

fabaff reviewed (commented):
Just some quick comments to get started.

markusressel committed:
reordered imports
added config defaults
check if platform is available
changed docstring

markusressel committed:
lint fix

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,50 @@
+Support for XS1 sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/xs1/
```
See above.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,50 @@
+
+from homeassistant.helpers.entity import Entity
+
+from homeassistant.components.xs1 import DOMAIN, SENSORS, XS1DeviceEntity
```
See above about domain.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,50 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_devices,
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,50 @@
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the sensor platform."""
+    _LOGGER.debug("initializing XS1 Sensor")
```
Remove this.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,50 @@
+
+    sensors = hass.data[DOMAIN][SENSORS]
+
+    _LOGGER.debug("Adding Sensor devices...")
```
Remove this.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+Support for XS1 switches.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/xs1/
```
See above.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+from functools import partial
+import logging
+
+from homeassistant.components.xs1 import ACTUATORS, DOMAIN, XS1DeviceEntity
```
See above about domain.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_devices,
```
See above.

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the XS1 platform."""
+    _LOGGER.debug("initializing XS1 Switch")
```
Remove this.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+
+    async_add_devices(switch_entities)
+
+    _LOGGER.debug("Added Switches!")
```
Remove this.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+        """Return true if switch is on."""
+        return self.device.value() == 100
+
+    async def async_turn_on(self, **kwargs):
```
Use the sync api.

markusressel replied:
Ok

MartinHjelmare commented on homeassistant/components/switch/xs1.py:
```diff
@@ -0,0 +1,60 @@
+        """Turn the device on."""
+        await self.hass.async_add_executor_job(
+            partial(self.device.turn_on))
+        self.async_schedule_update_ha_state()
```
`async_update` will be called at the end of the service call. So updating state in the actuator method is usually not needed.

markusressel replied:
Ok

markusressel commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=80): cv.string,
+        vol.Optional(CONF_SSL, default=False): cv.boolean,
+        vol.Optional(CONF_USERNAME, default=None): cv.string,
```
Ok

markusressel commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+    hass.data[DOMAIN][ACTUATORS] = actuators
+    hass.data[DOMAIN][SENSORS] = sensors
+
+    _LOGGER.debug("loading components for XS1 platform...")
```
Ok

markusressel commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+    """Set up XS1 Component."""
+    _LOGGER.debug("Initializing XS1")
+
+    host = config[DOMAIN].get(CONF_HOST)
```
Ok

markusressel commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+
+    host = config[DOMAIN].get(CONF_HOST)
+    port = config[DOMAIN].get(CONF_PORT)
+    ssl = config[DOMAIN].get(CONF_SSL, False)
```
Yes

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+Support for XS1 climate devices.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/xs1/
```
I have not created documentation for the platform specific code though.
I guess I also have to create pages for them (even if they are very simple)?

MartinHjelmare replied:
Yes, for now, our standard is that each component and each platform should have their own pages in the docs.

markusressel replied:
Done

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+
+from homeassistant.components.climate import ATTR_TEMPERATURE, ClimateDevice
+from homeassistant.components.xs1 import (
+    ACTUATORS, DOMAIN, SENSORS, XS1DeviceEntity)
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_devices,
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the XS1 platform."""
+    _LOGGER.debug("initializing XS1 Thermostat")
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+
+    async_add_devices(thermostat_entities)
+
+    _LOGGER.debug("Added Thermostats!")
```
Ok

markusressel committed:
review fixes

markusressel committed:
isort

markusressel committed:
import fix

markusressel commented on homeassistant/components/xs1.py:
```diff
@@ -0,0 +1,117 @@
+            user=user,
+            password=password)
+    except ConnectionError:
+        raise PlatformNotReady
```
So uh, what should I do then? Just remove it?

markusressel replied:
ok

markusressel committed:
review fix

markusressel committed:
review fix

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+        super().__init__(device)
+        self.sensor = sensor
+
+        self._min_temp = min_temp
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+    @property
+    def min_temp(self):
+        """Return the minimum temperature."""
+        return convert_temperature(
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+    @property
+    def max_temp(self):
+        """Return the maximum temperature."""
+        return convert_temperature(
```
Ok

markusressel commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,118 @@
+            self.unit_of_measurement
+        )
+
+    async def async_set_temperature(self, **kwargs):
```
Ok

markusressel committed:
review fix

markusressel committed:
removed unused imports

markusressel committed:
lint fix

MartinHjelmare commented:
To remove WIP status, please rebase and reword the commit message to remove WIP. I think it's only the first commit message that has wip in it.

markusressel committed:
lint fix

MartinHjelmare commented on homeassistant/components/climate/xs1.py:
```diff
@@ -0,0 +1,103 @@
+    """Set up the XS1 thermostat platform."""
+    from xs1_api_client.api_constants import ActuatorType
+
+    actuators = hass.data[COMPONENT_DOMAIN][ACTUATORS]
```
I understand now that the climate platform is not something the gateway actually supports.

We should remove this platform. It's already possible to use the generic thermostat climate platform for this purpose.

markusressel replied:
What do you mean by "dosen't support"? o.O I used this climate platform for years without any problem (in my own custom component)

MartinHjelmare replied:
It's just individual switches and sensors, correct? Not an actual thermostat?

MartinHjelmare replied:
The problem is that we shouldn't have more than one platform for the same api representation. If they are are switches and sensors they should be represented by those platforms. If they are part of a thermostat we should only use the climate platform. We have to decide what they are.

Will the switches and sensors always be hooked up to a thermostat or is that just one of many use cases? If thats the case we should let them be switches and sensors and not make a climate platform.

markusressel replied:
But it's not just a switch, you can set the temperature directly on the thermostat using the `actuator` and read the room temperature using the `sensor`. As I understand it this can't be reproduced by using a generic thermostat can it? An `actuator`  is not only a switch, it can be of many different types.

MartinHjelmare replied:
I read the code again now. Am I right that the same switch won't show up under two different platforms? It looks like the thermostat switches will only be connected to the climate platform. That's good.

How about the sensors? Will they show up under more than one platform?

markusressel replied:
I don't know why you call it a switch but yes, the "switch" **wont** show under two different platforms, the sensor in it's current implementation **will** show up as normal temperature sensor too.

MartinHjelmare replied:
I meant actuator, sorry.

Could we limit the thermostat sensor to the climate platform?

markusressel replied:
Any thoughts on this?
Imo it works out as it is. When the user does not know about the naming convention he will just get a temperature sensor and a climate device. If he knows about it they will be combined in a single sensor (and only for climate devices). I don't think this would mess up anything.

MartinHjelmare replied:
But all device sensors will always be set up as sensor entities. Can we try decide already in the component, what devices will be set up under what platforms? Ie, move the check for name to the component, and split sensors between climate and sensor platforms, if there's a name convention.

markusressel replied:
I added the same check to the sensor platform to prevent the sensor entity from showing up in two platforms if it used in a climate entity.

markusressel commented:
Well on the XS1 Gateway the actuator is a separate device from the sensor but they belong to the same physical device (which is a thermostat).

markusressel committed:
climate fix
exclude sensors that will be used in climate component from default sensor category

markusressel committed:
.coveragerc fix

markusressel committed:
lint fix

MartinHjelmare commented on homeassistant/components/sensor/xs1.py:
```diff
@@ -0,0 +1,57 @@
+    sensor_entities = []
+    for sensor in sensors:
+        belongs_to_climate_actuator = False
+        for actuator in actuators:
```
Move this to a separate iteration and save the result. That will save iterations.
```py
saved_actuators = {
    actuator.name(): actuator.type()
    for actuator in actuators}
for sensor in sensors:
     if saved_actuators.get(
            sensor.name()) == ActuatorType.TEMPERATURE:
        continue
    sensor_entities.append(XS1Sensor(sensor))
```

markusressel replied:
It doesn't work like that though, because the actuator and sensor names may not be equal (even though it should be possible in theory **the gateway does not allow this**)
Because of this I used the "in" notation.

markusressel replied:
The worst case of this loop is 64 * 64 which to be fair is quite a lot, but since only configured and active devices (on the gateway side) are considered in this loop I highly doubt this number will actually be reached by anyone. Additionally this check will only happen once when hass is started (at least in the current implementation).

MartinHjelmare replied:
Ok, :+1:.

MartinHjelmare commented on .coveragerc:
```diff
@@ -654,6 +657,7 @@ omit =
     homeassistant/components/wirelesstag/*
     homeassistant/components/xiaomi_aqara/*
     homeassistant/components/xiaomi_miio/*
+    homeassistant/components/xs1.py
```
Please move the component to a package, ie its own directory under components. That's our latest way of structuring things.

markusressel replied:
done

MartinHjelmare commented on .coveragerc:
```diff
@@ -81,6 +81,7 @@ omit =
     homeassistant/components/climate/sensibo.py
     homeassistant/components/climate/touchline.py
     homeassistant/components/climate/venstar.py
+    homeassistant/components/climate/xs1.py
```
Please move all the xs1 platforms under the xs1 component package.

markusressel replied:
done

markusressel committed:
moved platform to it's own package

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can be merged when build passes.

markusressel commented:
> Can be merged when build passes.

I think it did :tada:
I updated the documentation branch to reflect the next major Hass release (0.88).
If this is too late/early please let me know.
